cirq-core/cirq/transformers/target_gatesets/sqrt_iswap_gateset.py

```diff
@@ -0,0 +1,281 @@
+    """Target gateset containing √iSWAP + single qubit rotations + Measurement gates."""
+
+    def __init__(
+        self,
+        *,
+        atol: float = 1e-8,
+        required_sqrt_iswap_count: Optional[int] = None,
+        use_sqrt_iswap_inv: bool = False,
+    ):
+        """Initializes `cirq.SqrtIswapTargetGateset`
+
+        Args:
+            atol: A limit on the amount of absolute error introduced by the decomposition.
+            required_sqrt_iswap_count: When specified, the `decompose_to_target_gateset` will
+                decompose each operation into exactly this many sqrt-iSWAP gates even if fewer is
+                possible (maximum 3). A ValueError will be raised if this number is 2 or lower and
+                synthesis of the operation requires more.
+            use_sqrt_iswap_inv: If True, `cirq.SQRT_ISWAP_INV` is used as part of the gateset,
+                instead of `cirq.SQRT_ISWAP`.
+
+        Raises:
+            ValueError: If `required_sqrt_iswap_count` is specified and is not 0, 1, 2, or 3.
+        """
+        if required_sqrt_iswap_count is not None and not 0 <= required_sqrt_iswap_count <= 3:
+            raise ValueError('the argument `required_sqrt_iswap_count` must be 0, 1, 2, or 3.')
+        super().__init__(
+            ops.SQRT_ISWAP_INV if use_sqrt_iswap_inv else ops.SQRT_ISWAP,
+            ops.MeasurementGate,
+            ops.AnyUnitaryGateFamily(1),
+            name='SqrtIswapInvTargetGateset' if use_sqrt_iswap_inv else 'SqrtIswapTargetGateset',
+        )
+        self.atol = atol
+        self.required_sqrt_iswap_count = required_sqrt_iswap_count
+        self.use_sqrt_iswap_inv = use_sqrt_iswap_inv
+
+    def _decompose_two_qubit_operation(self, op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
+        if protocols.has_unitary(op):
+            return two_qubit_to_sqrt_iswap.two_qubit_matrix_to_sqrt_iswap_operations(
+                op.qubits[0],
+                op.qubits[1],
+                protocols.unitary(op),
+                required_sqrt_iswap_count=self.required_sqrt_iswap_count,
+                use_sqrt_iswap_inv=self.use_sqrt_iswap_inv,
+                atol=self.atol,
+                check_preconditions=False,
+                clean_operations=True,
+            )
+        gate = op.gate
+        q0, q1 = op.qubits
+
+        if isinstance(gate, ops.CZPowGate):
+            return _cphase_symbols_to_sqrt_iswap(q0, q1, gate.exponent, self.use_sqrt_iswap_inv)
+        if isinstance(gate, ops.SwapPowGate):
+            return _swap_symbols_to_sqrt_iswap(q0, q1, gate.exponent, self.use_sqrt_iswap_inv)
+        if isinstance(gate, ops.ISwapPowGate):
+            return _iswap_symbols_to_sqrt_iswap(q0, q1, gate.exponent, self.use_sqrt_iswap_inv)
+        if isinstance(gate, ops.FSimGate):
+            return _fsim_symbols_to_sqrt_iswap(
+                q0, q1, gate.theta, gate.phi, self.use_sqrt_iswap_inv
+            )
```

Comment:
Why are these extra cases needed here ? It looks like two_qubit_to_sqrt_iswap handles a lot of those (or could be made to much more cleanly there).

Reply:
These are cases for compiling parameterized gates. `two_qubit_matrix_to_sqrt_iswap_operations` does analytical decomposition of a two qubit matrix and hence will not work for parameterized gates.